gammapy/datasets/simulate.py

```diff
@@ -576,3 +578,99 @@ def run(self, dataset, observation=None):
+        self,
+        energy_axis_true,
+        energy_axis,
+        migra_axis,
+        map_width,
+        map_bin_size=0.02 * u.deg,
+        random_state="random-seed",
+    ):
+        self.random_state = get_random_state(random_state)
+        self.energy_axis_true = energy_axis_true
+        self.energy_axis = energy_axis
+        self.migra_axis = migra_axis
+        self.map_width = map_width
+        self.bin_size = map_bin_size
+
+    def run(self, observation: Observation, models=None):
+        """Run the event sampler, applying IRF corrections.
+
+        Parameters
+        ----------
+        observation : `~gammapy.data.Observation`
+            Observation to be simulated.
+        models : `~gammapy.modeling.Models`
+
+        Returns
+        -------
+        events : `~gammapy.data.EventList`
+            Event list.
+        """
+        if models is None:
+            models = Models()
+
+        pointing_icrs = observation.pointing.get_icrs(
+            observation.obstime,
+            observation.observatory_earth_location,
+        )
+        geom = WcsGeom.create(
+            skydir=pointing_icrs,
+            width=(self.map_width.to_value(u.deg), self.map_width.to_value(u.deg)),
+            binsz=self.bin_size.to_value(u.deg),
+            frame="icrs",
+            axes=[self.energy_axis],
+        )
+
+        dataset = MapDataset.create(
+            geom,
+            energy_axis_true=self.energy_axis_true,
+            migra_axis=self.migra_axis,
+            name="simulated-dataset",
+        )
+
+        components = ["exposure"]
+        if observation.edisp is not None:
+            components.append("edisp")
+        if observation.bkg is not None:
+            components.append("background")
+        if observation.psf is not None:
+            components.append("psf")
+
+        maker = MapDatasetMaker(selection=components)
+        dataset = maker.run(dataset, observation)
+        models.append(FoVBackgroundModel(dataset_name="simulated-dataset"))
+        dataset.models = models
+
+        events = None
+        if len(models) > 1:
+            events = self.sample_sources(dataset)
+
+            if len(events.table) > 0:
+                if dataset.psf is not None:
+                    events = self.sample_psf(dataset.psf, events)
+
+                if dataset.edisp is not None:
+                    events = self.sample_edisp(dataset.edisp, events)
+
+        if dataset.background is not None:
+            events_bkg = self.sample_background(dataset)
+            if events is not None:
+                events = EventList.from_stack([events_bkg, events])
+            else:
+                events = events_bkg
+
+        geom = dataset._geom
+        selection = geom.contains(events.map_coord(geom))
+        events = events.select_row_subset(selection)
+        events.table.sort("TIME")
+        events.table["EVENT_ID"] = np.arange(len(events.table))
+        observation._events = events
```

Comment:
Note: better return a copy with events set instead of mutating the input observation